File: sessions/week-01-two-pointers/algorithms/javascript/findMiddleOfLinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

function findMiddle(head) {
    let slow = head, fast = head;
    while (fast !== null && fast.next !== null) {
        slow = slow.next;
        fast = fast.next.next;
    }
    return slow;
}

// Example usage
function buildLinkedList(values) {
    const head = new Node(values[0]);
    let current = head;
    for (let i = 1; i < values.length; i++) {
        current.next = new Node(values[i]);
        current = current.next;
    }
    return head;
}

const head = buildLinkedList([1, 2, 3, 4, 5]);
const middle = findMiddle(head);
console.log("Middle node value:", middle.value);

File: sessions/week-01-two-pointers/algorithms/python/find_middle_of_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

def find_middle(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    return slow

# Example usage
def build_linked_list(values):
    head = Node(values[0])
    current = head
    for val in values[1:]:
        current.next = Node(val)
        current = current.next
    return head

head = build_linked_list([1, 2, 3, 4, 5])
middle = find_middle(head)
print("Middle node value:", middle.value)
